Add string template support for all rules

akirk opened the pull request:
Fixes #13

akirk committed:
Add string template support for all rules

aduth commented:
I worry about many of the conditions and checks in these rules now, since `getI18nStringFromNode` is a little too forgiving to its input. For non-string nodes, it will always return a string so [checks like this one](https://github.com/Automattic/eslint-plugin-wpcalypso/blob/b2bc38112072ce9f86153ffb35e800a2a441c500/lib/rules/i18n-mismatched-placeholders.js#L43-L46) are at best redundant and at worst not catching errors.

Do you (or @deBhal) think it might make sense for `getI18nStringFromNode` to return something different like `null` if the input node cannot be reasonably inferred as a string?

aduth commented:
Demonstrating current behavior with a test case in `test/lib/util/get-i18n-string-from-node.js`:

``` js
it( 'should return ??? for non-string input', () => {
    assert.equal( '', getTranslatableStringsFromCode( 'foo()' ) );
} );
```

Commit:
Rework getI18nStringFromNode() to return false for non-string nodes

Commit:
Check for string before calling indexOf() in i18n-ellipsis

deBhal commented:
@aduth I agree.  

Otherwise, I like this change, so I've reworked getI18nStringFromNode() to return `false` where nodes are invalid, and added a couple of tests with non-strings.  I had to add one check in i18n-ellipsis to avoid `false.indexOf()`, but the rest looks good.

👍  from me, if someone could just review my  2 commits.

akirk commented:
Nice! One thing remains for me though. Why `i18n` in `getI18nStringFromNode`? What about `extractStringFromNode`?

deBhal commented:
It's a shorthand for "assuming that template literals are boring".  `extractStringFromNode()` should do something sensible with `translate( `I have ${numberOfHats} hats` )`, but we return `false`.

It's not really `i18n`, but I feel like that gives us licence to make arbitrary limitations.  I'm open to a better name if you've got one.

akirk committed:
Rename getI18nStringFromNode to getTextContentFromNode

akirk committed:
Always return TemplateLiteral contents

akirk commented:
> extractStringFromNode() should do something sensible with translate( `I have ${numberOfHats} hats` ), but we return false.

I have now renamed it to `getTextContentFromNode` and changed the behavior so that it doesn't `return false` when it contains expressions which is [caught by this already](https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/lib/rules/i18n-no-variables.js#L36).

deBhal commented:
👍  I'm good with the functionality & name change, and all the tests pass.  🚢

akirk committed:
Update README and CHANGELOG

aduth commented on lib/rules/i18n-ellipsis.js:
```diff
@@ -24,16 +25,13 @@ var rule = module.exports = function( context ) {
-				}
-
-				if ( -1 !== arg.value.indexOf( '...' ) ) {
+				let argumentString = getTextContentFromNode( arg );
```
I found CircleCI had been silently failing on tests and this had failed to be flagged as an issue. Our `package.json` Node supports at the time of this pull request was `0.10.0`, where `let` is not a valid keyword.
- Fixed silent failing tests in d06a13fb4e8650994e1182c3c39c4a346c6f025e
- Replaced `let` with `var` in 6e2a654d3a7f101a78caa6e704eb2d4403a2fc48
- Bumped Node minimum to 0.12.0 in 8698df573578800ed388191af6f8fa0ee4481f41 
  - Because `0.10.0` doesn't satisfy our [`requireindex` dependency's own minimum](https://github.com/stephenhandley/requireindex/blob/b6cb37fc11923e6bc36a3db29b70e90b82c91fbb/package.json#L26))

deBhal replied:
Wow, thanks for noticing & fixing!

It's pretty weird that's a silent failure - I would have thought not supporting let would lead to a syntax error.  Did you happen to see why CircleCI didn't crash and burn?

Ah, looked at d06a13f and now I get it :/

Still, great catch!

aduth replied:
@deBhal : Yeah, if you look at the last build for this branch, you'll see that while green, the `npm test` step did produce an error:

https://circleci.com/gh/Automattic/eslint-plugin-wpcalypso/36
